### DISCUSSION TITLE
Mealie with LDAP (using lldap) not adding user...

### DISCUSSION THREAD

**Discussion by Entheologian** [mealie-recipes/mealie#2312] [Q&A]
Hi,

I am using Mealie in Docker and I tried using the config mentioned here, without any success yet:
https://github.com/lldap/lldap/discussions/291#discussioncomment-5496297

When I try to login it says "Invalid Credentials" and the mealie logs show this:
> mealie-api         | 192.168.112.3:57872 - "POST /api/auth/token HTTP/1.1" 401

The comment linked above mentioned this mealie issue, which I don't know if it's related or not:
https://github.com/hay-kot/mealie/issues/1775

In my case, I am trying to use ldaps but I have tried without the SSL (and using non-ssl port of course) and it doesn't seem to be hitting the ldap server, according to what I see in it's logs.

Has anyone been able go get mealie authenticating to LDAP, specifically with [lldap](https://github.com/lldap/lldap)?

I have these LDAP options set in my docker-compose.yml file:
* LDAP_AUTH_ENABLED=true
* LDAP_SERVER_URL="ldaps://192.168.0.210:636/"
* LDAP_BIND_TEMPLATE="uid={},ou=people,dc=domain,dc=net"
* LDAP_TLS_INSECURE=true #tried both true/false
* LDAP_ENABLE_STARTTLS=true #tried both true/false
* LDAP_BASE_DN="dc=domain,dc=net"
* LDAP_QUERY_PASSWORD=mypassword

**Comment by cmintey** [comment]
What version are you running? LDAP is fixed in the latest nightly version and it comes with some new [environment variables](https://nightly.mealie.io/documentation/getting-started/installation/backend-config/)

**Reply by Entheologian** [reply]
Is this the version you were asking about? 
mealie:frontend-v1.0.0beta-4

**Reply by Entheologian** [reply]
I re-tried with latest nightly builds and at best, this is the error I see with these config options (when trying without ldaps). 

Am I doing something incorrectly? I also tried with LDAP_ADMIN_FILTER / LDAP_USER_FILTER, but same error:
```
mealie-api         |     "LDAP_AUTH_ENABLED": true,
mealie-api         |     "LDAP_SERVER_URL": "\"ldap://192.168.0.210:389\"",
mealie-api         |     "LDAP_TLS_INSECURE": false,
mealie-api         |     "LDAP_TLS_CACERTFILE": null,
mealie-api         |     "LDAP_ENABLE_STARTTLS": false,
mealie-api         |     "LDAP_BASE_DN": "dc=domain,dc=net",
mealie-api         |     "LDAP_QUERY_BIND": "cn=admin,ou=people,dc=domain,dc=net",
mealie-api         |     "LDAP_QUERY_PASSWORD": "password",
mealie-api         |     "LDAP_USER_FILTER": null,
mealie-api         |     "LDAP_ADMIN_FILTER": null,
mealie-api         |     "LDAP_ID_ATTRIBUTE": "uid",
mealie-api         |     "LDAP_MAIL_ATTRIBUTE": "mail",
mealie-api         |     "LDAP_NAME_ATTRIBUTE": "name",
mealie-api         |     "TESTING": false
mealie-api         | }
mealie-api         | INFO:     Application startup complete.
mealie-api         | INFO:     Uvicorn running on http://0.0.0.0:9000 (Press CTRL+C to quit)
mealie-api         | INFO:     172.27.0.3:60598 - "POST /api/auth/token HTTP/1.1" 500 Internal Server Error
mealie-api         | ERROR:    Exception in ASGI application
mealie-api         | Traceback (most recent call last):
mealie-api         |   File "/opt/pysetup/.venv/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 436, in run_asgi
mealie-api         |     result = await app(  # type: ignore[func-returns-value]
mealie-api         |   File "/opt/pysetup/.venv/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 78, in __call__
mealie-api         |     return await self.app(scope, receive, send)
mealie-api         |   File "/opt/pysetup/.venv/lib/python3.10/site-packages/fastapi/applications.py", line 276, in __call__
mealie-api         |     await super().__call__(scope, receive, send)
mealie-api         |   File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/applications.py", line 122, in __call__
mealie-api         |     await self.middleware_stack(scope, receive, send)
mealie-api         |   File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/middleware/errors.py", line 184, in __call__
mealie-api         |     raise exc
mealie-api         |   File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/middleware/errors.py", line 162, in __call__
mealie-api         |     await self.app(scope, receive, _send)
mealie-api         |   File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/middleware/gzip.py", line 24, in __call__
mealie-api         |     await responder(scope, receive, send)
mealie-api         |   File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/middleware/gzip.py", line 44, in __call__
mealie-api         |     await self.app(scope, receive, self.send_with_gzip)
mealie-api         |   File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 79, in __call__
mealie-api         |     raise exc
mealie-api         |   File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 68, in __call__
mealie-api         |     await self.app(scope, receive, sender)
mealie-api         |   File "/opt/pysetup/.venv/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
mealie-api         |     raise e
mealie-api         |   File "/opt/pysetup/.venv/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
mealie-api         |     await self.app(scope, receive, send)
mealie-api         |   File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/routing.py", line 718, in __call__
mealie-api         |     await route.handle(scope, receive, send)
mealie-api         |   File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/routing.py", line 276, in handle
mealie-api         |     await self.app(scope, receive, send)
mealie-api         |   File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/routing.py", line 66, in app
mealie-api         |     response = await func(request)
mealie-api         |   File "/opt/pysetup/.venv/lib/python3.10/site-packages/fastapi/routing.py", line 237, in app
mealie-api         |     raw_response = await run_endpoint_function(
mealie-api         |   File "/opt/pysetup/.venv/lib/python3.10/site-packages/fastapi/routing.py", line 165, in run_endpoint_function
mealie-api         |     return await run_in_threadpool(dependant.call, **values)
mealie-api         |   File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/concurrency.py", line 41, in run_in_threadpool
mealie-api         |     return await anyio.to_thread.run_sync(func, *args)
mealie-api         |   File "/opt/pysetup/.venv/lib/python3.10/site-packages/anyio/to_thread.py", line 31, in run_sync
mealie-api         |     return await get_asynclib().run_sync_in_worker_thread(
mealie-api         |   File "/opt/pysetup/.venv/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 937, in run_sync_in_worker_thread
mealie-api         |     return await future
mealie-api         |   File "/opt/pysetup/.venv/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 867, in run
mealie-api         |     result = context.run(func, *args)
mealie-api         |   File "/app/mealie/routes/auth/auth.py", line 56, in get_token
mealie-api         |     user = authenticate_user(session, email, password)  # type: ignore
mealie-api         |   File "/app/mealie/core/security/security.py", line 168, in authenticate_user
mealie-api         |     return user_from_ldap(db, email, password)
mealie-api         |   File "/app/mealie/core/security/security.py", line 64, in user_from_ldap
mealie-api         |     conn = ldap.initialize(settings.LDAP_SERVER_URL)
mealie-api         |   File "/opt/pysetup/.venv/lib/python3.10/site-packages/ldap/functions.py", line 91, in initialize
mealie-api         |     return LDAPObject(
mealie-api         |   File "/opt/pysetup/.venv/lib/python3.10/site-packages/ldap/ldapobject.py", line 88, in __init__
mealie-api         |     self._l = ldap.functions._ldap_function_call(ldap._ldap_module_lock,_ldap.initialize,uri)
mealie-api         |   File "/opt/pysetup/.venv/lib/python3.10/site-packages/ldap/functions.py", line 52, in _ldap_function_call
mealie-api         |     result = func(*args,**kwargs)
mealie-api         | ldap.LDAPError: (0, 'Error')
```

**Reply by cmintey** [reply]
This is a different error. It looks like your ldap server cannot be reached, it fails at the initialize step

**Reply by cmintey** [reply]
FWIW, this is my configuration, using lldap:
```
LDAP_AUTH_ENABLED=true
LDAP_SERVER_URL=ldap://10.0.2.100:3890
LDAP_TLS_INSECURE=true
LDAP_BASE_DN=ou=people,dc=example,DC=com
LDAP_QUERY_BIND="uid=admin,ou=people,dc=example,dc=com"
LDAP_QUERY_PASSWORD="password"
LDAP_USER_FILTER="(&(objectClass=person)(memberOf=cn=mealie_user,ou=groups,dc=example,dc=com))"
LDAP_ADMIN_FILTER="memberOf=cn=mealie_admin,ou=groups,dc=example,dc=com"
LDAP_ID_ATTRIBUTE=uid
LDAP_NAME_ATTRIBUTE=cn
LDAP_MAIL_ATTRIBUTE=mail
```

The admin and user filters are optional, they just narrow down to specific users who can access and who are admin

**Reply by cmintey** [reply]
Taking a closer look at your config, I think this might be the issue
> "LDAP_SERVER_URL": "\"ldap://192.168.0.210:389\""

The ldap url is double quoted. Try removing the quotes around the url on your environment variable

**Reply by Entheologian** [reply]
Awesome, the double quotes was the problem.

After I removed them from the docker-compose.yml file it queried the LDAP but couldn't find the base DN which gave another error, so I removed the double quotes from my LDAP_BASE_DN line (seen in my initial post above) and that too fixed it.

Using your example, this didn't work for me:
LDAP_BASE_DN="ou=people,dc=example,DC=com"
LDAP_SERVER_URL="ldap://10.0.2.100:3890"

But this way did:
LDAP_BASE_DN=ou=people,dc=example,DC=com
LDAP_SERVER_URL=ldap://10.0.2.100:3890

Thank you Carter! =)

**Reply by Entheologian** [reply]
I do have one simple question remaining...
From https://nightly.mealie.io/documentation/getting-started/usage/ldap/ it says:
> If the user already has an account in Mealie and wants to use their LDAP credentials instead, then you can go to the User Management page in the admin panel and change the "Authentication Backend" from Mealie to LDAP.

How do I access that admin panel to change that? I didn't see that LDAP option in the profile page.

**Reply by cmintey** [reply]
Click on **Settings** in the lower left corner, then click **Users**, and you should see a list of all your users. You can click on one of them and change the Auth method there

**Reply by Entheologian** [reply]
Oh interesting, on the left corner I only see Language and Light/Dark Mode. Is that where it should be?

**Reply by cmintey** [reply]
Yeah it should be there if you're logged in as an admin. Do you have a local admin account to log in with?

**Reply by Entheologian** [reply]
Ahh, an admin account would explain this issue... I don't have one and didn't even know it had that option. I didn't see it in the documentation, unless I missed it?

Can you please point me to where/how I can read how to setup the admin account?

**Reply by cmintey** [reply]
The default user is an admin. 
```
Username: changeme@email.com
Password: MyPassword
```

If you've deleted this user, then one way you can add yourself as admin (since you're using LDAP), is to add in an `LDAP_ADMIN_FILTER` as shown in the documentation. You can create a mealie admin group (or use whatever existing group) in lldap and assign yourself.

**Reply by Entheologian** [reply]
Ahhh, I will do it that way. Thanks for your time!

**Reply by cmintey** [reply]
No problem, glad to help
